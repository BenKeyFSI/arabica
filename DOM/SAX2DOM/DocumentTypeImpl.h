#ifndef JEZUK_SAX2DOM_DOCUMENTTYPEIMPL_H
#define JEZUK_SAX2DOM_DOCUMENTTYPEIMPL_H

#include <DOM/Simple/DocumentTypeImpl.h>

namespace SAX2DOM
{

template<class stringT, class string_adaptorT>
class DocumentType : public SimpleDOM::DocumentTypeImpl<stringT, string_adaptorT>
{
    typedef SimpleDOM::DocumentTypeImpl<stringT, string_adaptorT> BaseDT;
  public:
    DocumentType(const stringT& qualifiedName,
                 const stringT& publicId,
                 const stringT& systemId) :
      BaseDT(qualifiedName, publicId, systemId)
    {
    } // DocumentType

    virtual ~DocumentType() { }

    //////////////////////////////////////////////////////
    // DOM::DOMImplementation extensions
    void addEntity(SimpleDOM::EntityImpl<stringT, string_adaptorT>* entity)
    {
      BaseDT::addEntity(entity);
    } // addEntity

    void addNotation(SimpleDOM::NotationImpl<stringT, string_adaptorT>* notation)
    {
      BaseDT::addNotation(notation);
    } // addNotation

    void addElement(const stringT& name)
    {
      BaseDT::addElement(new SimpleDOM::ElementImpl<stringT, string_adaptorT>(BaseDT::ownerDoc_, name));
    } // addElement

    void addDefaultAttr(const stringT& elementName, const stringT& attributeName, const stringT& attributeValue)
    {
      SimpleDOM::ElementImpl<stringT, string_adaptorT>* elem = dynamic_cast<SimpleDOM::ElementImpl<stringT, string_adaptorT>*>(BaseDT::getElements()->getNamedItem(elementName));
      if(elem == 0)
      {
        addElement(elementName);
        elem = dynamic_cast<SimpleDOM::ElementImpl<stringT, string_adaptorT>*>(BaseDT::getElements()->getNamedItem(elementName));
      }
      SimpleDOM::AttrImpl<stringT, string_adaptorT>* attr = new SimpleDOM::AttrImpl<stringT, string_adaptorT>(BaseDT::ownerDoc_, attributeName);
      attr->setValue(attributeValue);
      attr->setSpecified(false);
      elem->setAttributeNode(attr);
    } // addDefaultAttr

    void addElementId(const stringT& elementName)
    {
      BaseDT::addElementId(elementName);
    } // addElementId
}; // class DOMImplementation

} // namespace SAX2DOM

#endif
// end of file
