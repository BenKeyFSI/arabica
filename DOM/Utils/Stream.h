#ifndef ARABICA_DOM_UTILS_STREAM_H
#define ARABICA_DOM_UTILS_STREAM_H
///////////////////////////////////////////////////////////////////////
//
// DOM/Utils/Stream.h
// 
// Written by Jez Higgins <jez@jezuk.co.uk>
// Copyright 2003 Jez UK Ltd, http://www.jezuk.co.uk
//
// Provides streaming operator<< for DOM::Nodes.  Fully parameterised so
// will work with wide and narrow char types, so long as an operator<< 
// exists for the DOM::Node's string_type.
//
// e.g.
//  SAX2DOM::Parser<std::wstring> domParser;
//  SAX::wInputSource is;
//  ... load instance from somewhere ...
//  domParser.parse(is);
//  DOM::Document<std::wstring> doc = domParser.getDocument();
//  ... manipulate doc ...
//  std::wcout << doc;  
//
///////////////////////////////////////////////////////////////////////
// $Id$
///////////////////////////////////////////////////////////////////////

#include <iostream>
#include <algorithm>
#include <XML/UnicodeCharacters.h>

namespace DOM
{

  namespace StreamImpl
  {
    template<typename char_type, typename traits_type>
    class escaper
    {
      private:
        typedef char_type charT;
        typedef traits_type traitsT;
        typedef std::basic_ostream<charT, traitsT> ostreamT;
        typedef Arabica::Unicode<charT> UnicodeT;

      public:
        escaper(ostreamT& stream) : stream_(stream) { }
        void operator()(charT ch)
        {
	        if(ch == UnicodeT::LESS_THAN_SIGN)
	        {
            stream_ << UnicodeT::AMPERSAND
                      << UnicodeT::LOWERCASE_L
                      << UnicodeT::LOWERCASE_T
                      << UnicodeT::SEMI_COLON;
	          return;
          } // if(ch == UnicodeT::LESS_THAN_SIGN)
          if(ch == UnicodeT::GREATER_THAN_SIGN)
	        {
            stream_ << UnicodeT::AMPERSAND
                    << UnicodeT::LOWERCASE_G
                    << UnicodeT::LOWERCASE_T
                    << UnicodeT::SEMI_COLON;
            return;
	        } // if(ch == UnicodeT::GREATER_THAN_SIGN)
          if(ch == UnicodeT::AMPERSAND)
	        {
            stream_ << UnicodeT::AMPERSAND
                    << UnicodeT::LOWERCASE_A
                    << UnicodeT::LOWERCASE_M
                    << UnicodeT::LOWERCASE_P
                    << UnicodeT::SEMI_COLON;
	          return;
          } // if(ch == case UnicodeT::AMPERSAND)
          if(ch == UnicodeT::QUOTATION_MARK)
	        {
            stream_ << UnicodeT::AMPERSAND
                    << UnicodeT::LOWERCASE_Q
                    << UnicodeT::LOWERCASE_U
                    << UnicodeT::LOWERCASE_O
                    << UnicodeT::LOWERCASE_T
                    << UnicodeT::SEMI_COLON;
            return;
	        } // if(ch == UnicodeT::QUOTATION_MARK)
          stream_ << ch;
        } // operator()

      private:
        ostreamT& stream_;
    }; // escaper

    template<class stringT, class charT, class traitsT>
    void streamChildren(std::basic_ostream<charT, traitsT>& stream, DOM::Node<stringT>& node)
    {
      DOM::Node<stringT> child = node.getFirstChild();
      while(child != 0)
      {
        stream << child;
        child = child.getNextSibling();
      } // while 
    } // streamChildren
  
  } // namespace StreamImpl

template<class stringT, class charT, class traitsT>
std::basic_ostream<charT, traitsT>&
operator<<(std::basic_ostream<charT, traitsT>& stream,
           DOM::Node<stringT>& node)
{
  typedef Arabica::Unicode<charT> UnicodeT;

  switch(node.getNodeType())
  {
  case DOM::Node<stringT>::DOCUMENT_NODE:
    stream << UnicodeT::LESS_THAN_SIGN
           << UnicodeT::QUESTION_MARK
           << UnicodeT::LOWERCASE_X
           << UnicodeT::LOWERCASE_M
           << UnicodeT::LOWERCASE_L
           << UnicodeT::SPACE
           << UnicodeT::LOWERCASE_V
           << UnicodeT::LOWERCASE_E
           << UnicodeT::LOWERCASE_R
           << UnicodeT::LOWERCASE_S
           << UnicodeT::LOWERCASE_I
           << UnicodeT::LOWERCASE_O
           << UnicodeT::LOWERCASE_N
           << UnicodeT::EQUALS_SIGN
           << UnicodeT::QUOTATION_MARK
           << UnicodeT::NUMBER_1
           << UnicodeT::FULL_STOP
           << UnicodeT::NUMBER_0
           << UnicodeT::QUOTATION_MARK
           << UnicodeT::QUESTION_MARK
           << UnicodeT::GREATER_THAN_SIGN
           << std::endl;
    StreamImpl::streamChildren(stream, node);
    break;
  case DOM::Node<stringT>::ELEMENT_NODE:
    {
      stream << UnicodeT::LESS_THAN_SIGN << node.getNodeName();
      DOM::NamedNodeMap<stringT> attrs = node.getAttributes();
      for(unsigned int a = 0; a < attrs.getLength(); ++a)
      {
        DOM::Node<stringT> attr = attrs.item(a);
        stream << UnicodeT::SPACE 
               << attr.getNodeName()
               << UnicodeT::EQUALS_SIGN
               << UnicodeT::QUOTATION_MARK;
        stringT value = attr.getNodeValue();
        std::for_each(value.begin(), value.end(), StreamImpl::escaper<charT, traitsT>(stream));
        stream << UnicodeT::QUOTATION_MARK;
      }
      stream << UnicodeT::GREATER_THAN_SIGN;
      StreamImpl::streamChildren(stream, node);
      stream << UnicodeT::LESS_THAN_SIGN << UnicodeT::SLASH << node.getNodeName() << UnicodeT::GREATER_THAN_SIGN;
    }
    break;
  case DOM::Node<stringT>::TEXT_NODE:
    {
      stringT value = node.getNodeValue();
      std::for_each(value.begin(), value.end(), StreamImpl::escaper<charT, traitsT>(stream));
    }
    break;
  case DOM::Node<stringT>::ENTITY_REFERENCE_NODE:
    stream << UnicodeT::AMPERSAND
           << node.getNodeName()
           << UnicodeT::SEMI_COLON;
    break;
  case DOM::Node<stringT>::CDATA_SECTION_NODE:
    stream << UnicodeT::LESS_THAN_SIGN
           << UnicodeT::EXCLAMATION_MARK
           << UnicodeT::LEFT_SQUARE_BRACKET
           << UnicodeT::CAPITAL_C
           << UnicodeT::CAPITAL_D
           << UnicodeT::CAPITAL_A
           << UnicodeT::CAPITAL_T
           << UnicodeT::CAPITAL_A
           << UnicodeT::LEFT_SQUARE_BRACKET
           << node.getNodeValue()
           << UnicodeT::RIGHT_SQUARE_BRACKET
           << UnicodeT::RIGHT_SQUARE_BRACKET
           << UnicodeT::GREATER_THAN_SIGN;
    break;
  case DOM::Node<stringT>::PROCESSING_INSTRUCTION_NODE:
    stream << UnicodeT::LESS_THAN_SIGN
           << UnicodeT::QUESTION_MARK
           << node.getNodeName()
           << UnicodeT::SPACE
           << node.getNodeValue()
           << UnicodeT::QUESTION_MARK
           << UnicodeT::GREATER_THAN_SIGN;
    break;
  case DOM::Node<stringT>::COMMENT_NODE:
    stream << UnicodeT::LESS_THAN_SIGN
           << UnicodeT::EXCLAMATION_MARK
           << UnicodeT::HYPHEN_MINUS
           << UnicodeT::HYPHEN_MINUS
           << node.getNodeValue()
           << UnicodeT::HYPHEN_MINUS
           << UnicodeT::HYPHEN_MINUS
           << UnicodeT::GREATER_THAN_SIGN;
    break;
  default: // keep GCC quiet
	  break;
  } // switch

  return stream;
} // operator<<

} // namespace DOM
#endif
