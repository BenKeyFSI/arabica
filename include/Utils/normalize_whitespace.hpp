#ifndef ARABICA_UTILS_NORMALIZE_WHITESPACE_HPP
#define ARABICA_UTILS_NORMALIZE_WHITESPACE_HPP

#include <Utils/StringAdaptor.hpp>
#include <XML/XMLCharacterClasses.hpp>
#include <XML/UnicodeCharacters.hpp>

namespace Arabica
{

namespace string
{

template<class string_type, class string_adaptor>
string_type normalize_whitespace(const string_type& ch)
{
  std::string value = string_adaptor::asStdString(ch);
  std::string stripped = normalize_whitespace<std::string, Arabica::default_string_adaptor<std::string> >(value);
  return string_adaptor::construct_from_utf8(stripped.c_str());
} // normalize_whitespace

template<>
std::string normalize_whitespace<std::string, Arabica::default_string_adaptor<std::string> >(const std::string& ch)
{
  std::string value(ch);
  std::string::const_iterator i = value.begin(), ie = value.end();
  std::string::iterator p = value.begin(), pe = value.end();

  // string leading space
  while((i != ie) && (Arabica::XML::is_space(static_cast<char>(*i))))
    ++i;
    
  while(i != ie)
  {
    while((i != ie) && (!Arabica::XML::is_space(static_cast<char>(*i)))) 
      *p++ = *i++;
    while((i != ie) && (Arabica::XML::is_space(static_cast<char>(*i))))
      ++i;
    if(i != ie)
      *p++ = Arabica::Unicode<char>::SPACE;
  } // while ...
  if(p != pe)
    value.erase(p, pe);

  return value;
} // normalize_whitespace

} // namespace string

} // namespace Arabica
#endif

