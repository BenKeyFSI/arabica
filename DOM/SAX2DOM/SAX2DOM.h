#ifndef JEZUK_SAX2DOM_PARSER_H
#define JEZUK_SAX2DOM_PARSER_H

#include <SAX/XMLReader.h>
#include <SAX/ext/DefaultHandler2.h>
#include <SAX/helpers/AttributeTypes.h>
#include <DOM/Simple/DOMImplementation.h>
#include <DOM/Simple/NotationImpl.h>
#include <DOM/Simple/EntityImpl.h>
#include <DOM/Document.h>
#include <DOM/DOMException.h>
#include <DOM/SAX2DOM/DocumentTypeImpl.h>
#include <map>
#include <SAX/helpers/FeatureNames.h>
#include <SAX/helpers/PropertyNames.h>

namespace SAX2DOM
{

template<class stringT, 
         class string_adaptorT = SAX::default_string_adaptor<stringT>,
         class SAX_parser = SAX::XMLReader<stringT> >
class Parser : private SAX::basic_DefaultHandler2<stringT>
{
    typedef SAX::basic_EntityResolver<stringT> EntityResolverT;
    typedef SAX::basic_ErrorHandler<stringT> ErrorHandlerT;
    typedef SimpleDOM::EntityImpl<stringT, string_adaptorT> EntityT;
    typedef SimpleDOM::NotationImpl<stringT, string_adaptorT> NotationT;
    typedef SimpleDOM::ElementImpl<stringT, string_adaptorT> ElementT;
    typedef typename SAX::basic_ErrorHandler<stringT>::SAXParseExceptionT SAXParseExceptionT;

  public:
    Parser() :
        entityResolver_(0),
        errorHandler_(0)
    { 
    } // Parser

    void setEntityResolver(EntityResolverT& resolver) { entityResolver_ = &resolver; }
    EntityResolverT* getEntityResolver() const { return entityResolver_; }

    void setErrorHandler(ErrorHandlerT& handler) { errorHandler_ = &errorHandler; }
    ErrorHandlerT* getErrorHandler() const { return errorHandler_; }

    bool parse(const stringT& systemId)
    {
      SAX::basic_InputSource<stringT> is(systemId);
      return parse(is);
    } // loadDOM

    bool parse(SAX::basic_InputSource<stringT>& source)
    {
      SAX::FeatureNames<stringT> fNames;
      SAX::PropertyNames<stringT> pNames;
      
      DOM::DOMImplementation<stringT> di = SimpleDOM::DOMImplementation<stringT, string_adaptorT>::getDOMImplementation();
      document_ = di.createDocument(SA_.makeStringT(""), SA_.makeStringT(""), 0);
      currentNode_ = document_;
      inCDATA_ = false;
      inEntity_ = 0;

      SAX_parser parser;
		  parser.setContentHandler(*this);
		  parser.setErrorHandler(*this);
      if(entityResolver_)
        parser.setEntityResolver(*entityResolver_);
      try 
      {
        parser.setProperty(pNames.lexicalHandler, static_cast<SAX::basic_LexicalHandler<stringT>&>(*this));
        parser.setProperty(pNames.declHandler, static_cast<SAX::basic_DeclHandler<stringT>&>(*this));
      }
      catch(SAX::SAXException&)
      {
      } // catch
      parser.setFeature(fNames.namespaces, true);
      parser.setFeature(fNames.namespace_prefixes, true);
      //parser.setFeature(fNames.external_general, true);

      parser.parse(source);

      return (document_ != 0);
    } // loadDOM

    DOM::Document<stringT> getDocument() const
    {
      return document_;
    } // getDocument

    void reset()
    {
      currentNode_ = 0;
      document_ = 0;
    } // reset

  private:
    // no implementations
    Parser(const Parser&);
    bool operator==(const Parser&) const;
    Parser& operator=(const Parser&);

    // instance variables
    DOM::Document<stringT> document_;
    DocumentType<stringT, string_adaptorT >* documentType_;
    DOM::Node<stringT> currentNode_;
    DOM::Node<stringT> cachedCurrent_;

    std::string lastError_;
    bool inCDATA_;
    int inEntity_;
    string_adaptorT SA_;
    std::map<stringT, EntityT*> declaredEntities_;

    EntityResolverT* entityResolver_;
    ErrorHandlerT* errorHandler_;
    SAX::AttributeTypes<stringT, string_adaptorT> attributeTypes_;

  private:
    ///////////////////////////////////////////////////////////
    // ContentHandler
    virtual void endDocument()
    {
      currentNode_ = 0;
    } // endDocument

    virtual void startElement(const stringT& namespaceURI, const stringT& localName,
                              const stringT& qName, const SAX::basic_Attributes<stringT>& atts)
    {
      if(currentNode_ == 0)
        return;

      DOM::Element<stringT> elem = document_.createElementNS(namespaceURI, qName);
      currentNode_.appendChild(elem);

      // attributes here
      for(int i = 0; i < atts.getLength(); ++i)
        elem.setAttributeNS(atts.getURI(i), atts.getQName(i), atts.getValue(i));

      currentNode_ = elem;
    } // startElement

    virtual void endElement(const stringT& namespaceURI, const stringT& localName,
                            const stringT& qName)
    {
      if(currentNode_ == 0)
        return;

      currentNode_ = currentNode_.getParentNode();
    } // endElement

    virtual void characters(const stringT& ch)
    {
      if(currentNode_ == 0)
        return;

      if(!inCDATA_)
        currentNode_.appendChild(document_.createTextNode(ch));
      else
        currentNode_.appendChild(document_.createCDATASection(ch));
    } // characters

    virtual void processingInstruction(const stringT& target, const stringT& data)
    {
      if(currentNode_ == 0)
        return;

      currentNode_.appendChild(document_.createProcessingInstruction(target, data));
    } // processingInstruction

    virtual void skippedEntity(const stringT& name)
    {
      if(currentNode_ == 0)
        return;

      currentNode_.appendChild(document_.createEntityReference(name));
    } // skippedEntity

    ////////////////////////////////////////////////////
    // ErrorHandler
    virtual void warning(const SAXParseExceptionT& e) 
    { 
      if(errorHandler_)
        errorHandler_->warning(e);
    } // warning

    virtual void error(const SAXParseExceptionT& e) 
    {
      if(errorHandler_)
        errorHandler_->error(e);
      reset();
    } // error

    virtual void fatalError(const SAXParseExceptionT& e)
    {
      if(errorHandler_)
        errorHandler_->fatalError(e);
      reset();
    } // fatalError

    /////////////////////////////////////////////////////
    // LexicalHandler
    virtual void startDTD(const stringT& name, const stringT& publicId, const stringT& systemId)
    {
      documentType_ = new DocumentType<stringT, string_adaptorT >(name, publicId, systemId);
      document_.insertBefore(documentType_, 0);
    } // startDTD

    virtual void endDTD()
    {
      documentType_->setReadOnly(true);
    } // endDTD

    virtual void startEntity(const stringT& name)
    {
      if(currentNode_ == 0)
        return;

      if(++inEntity_ == 1)
      {
        cachedCurrent_ = currentNode_;
        currentNode_ = declaredEntities_[name];
        if(currentNode_ != 0 && currentNode_.hasChildNodes() == true) // already populated
          currentNode_ = 0;
      }
    } // startEntity

    virtual void endEntity(const stringT& name)
    {
      if(--inEntity_ == 0)
        currentNode_ = cachedCurrent_;

      currentNode_.appendChild(document_.createEntityReference(name));
    } // endEntity

    virtual void startCDATA()
    {
      inCDATA_ = true;
    } // startCDATA

    virtual void endCDATA()
    {
      inCDATA_ = false;
    } // endCDATA

    virtual void comment(const stringT& text)
    {
      if(currentNode_ == 0)
        return;

      currentNode_.appendChild(document_.createComment(text));
    } // comment

    //////////////////////////////////////////////////////////////////////
    // DeclHandler
    virtual void elementDecl(const stringT& name, const stringT& model)
    {
      documentType_->addElement(new ElementT(0, name));
    } // elementDecl

    virtual void attributeDecl(const stringT& elementName,
                               const stringT& attributeName,
                               const stringT& type,
                               const stringT& valueDefault,
                               const stringT& value)
    {
      if(!value.empty())
        documentType_->addDefaultAttr(elementName, attributeName, value);
      if(type == attributeTypes_.id)
        documentType_->addElementId(attributeName);
    } // attributeDecl

    virtual void internalEntityDecl(const stringT& name, const stringT& value)
    {
      EntityT* entity = new EntityT(0, name, SA_.makeStringT(""), SA_.makeStringT(""), SA_.makeStringT(""));
      declaredEntities_.insert(std::make_pair(name, entity));
      documentType_->addEntity(entity);
      currentNode_ = entity;
      currentNode_.appendChild(document_.createTextNode(value));
      currentNode_ = document_;
    } // internalEntityDecl

    virtual void externalEntityDecl(const stringT& name, const stringT& publicId, const stringT& systemId)
    {
      EntityT* entity = new EntityT(0, name, publicId, systemId, SA_.makeStringT(""));
      currentNode_ = entity;

      declaredEntities_.insert(std::make_pair(name, entity)); // we'll populate it later

      documentType_->addEntity(entity);
      currentNode_ = document_;
    } // externalEntityDecl

    /////////////////////////////////////////////////////////////////////////
    // DTDHandler
    virtual void notationDecl(const stringT& name, const stringT& publicId, const stringT& systemId)
    {
      documentType_->addNotation(new NotationT(0, name, publicId, systemId));
    } // notationDecl

    virtual void unparsedEntityDecl(const stringT& name, const stringT& publicId, const stringT& systemId, const stringT& notationName)
    {
      documentType_->addEntity(new EntityT(0, name, publicId, systemId, notationName));
    } // unparsedEntityDecl
}; // class Parser

} // namespace

#endif

