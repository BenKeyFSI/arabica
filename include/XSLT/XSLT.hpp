#ifndef ARABICA_XSLT_XSLT_HPP
#define ARABICA_XSLT_XSLT_HPP

#include <SAX/InputSource.hpp>
#include <SAX/Attributes.hpp>
#include <DOM/SAX2DOM/SAX2DOM.hpp>
#include <DOM/io/Stream.hpp>
#include <XSLT/impl/xslt_stylesheet_compiler.hpp>

#endif
