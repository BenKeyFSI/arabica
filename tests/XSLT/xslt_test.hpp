#ifndef XSLT_XSLT_TEST_HPP
#define XSLT_XSLT_TEST_HPP

#include <SAX/InputSource.hpp>
#include <DOM/SAX2DOM/SAX2DOM.hpp>
#include <DOM/Utils/Stream.hpp>
#include <XSLT/XSLT.hpp>
#include <Utils/normalize_whitespace.hpp>

#ifdef ARABICA_WINDOWS
const std::string PATH_PREFIX="../tests/XSLT/testsuite/TESTS/";
const std::string SEPERATOR = "/";
#else
#include "test_path.hpp"
const std::string PATH_PREFIX=test_path;
const std::string SEPERATOR = "/";
#endif
  
Arabica::DOM::Document<std::string> buildDOM(const std::string& filename)
{
  Arabica::SAX::InputSource is(filename);
  Arabica::SAX2DOM::Parser<std::string> parser;
  parser.parse(is);       

  Arabica::DOM::Document<std::string> d = parser.getDocument();
  if(d != 0)
    d.normalize();
  return d;
} // buildDOM

Arabica::XPath::NodeSet<std::string> selectNodes(const std::string& path, const Arabica::DOM::Node<std::string>& node)
{
  Arabica::XPath::XPath<std::string> xpath;
  return xpath.evaluate(path, node)->asNodeSet();
} // selectNodes

Arabica::DOM::Node<std::string> selectNode(const std::string& path, const Arabica::DOM::Node<std::string>& node)
{
  return selectNodes(path, node)[0];
} // selectNode

std::string selectString(const std::string& path, const Arabica::DOM::Node<std::string>& node)
{
  Arabica::XPath::XPath<std::string> xpath;
  return xpath.evaluate_expr(path, node)->asString();
} // selectString

std::string make_path(const std::string& path, const std::string& filename)
{
  return PATH_PREFIX + path + SEPERATOR + filename;
} // make_path

class SkipTest : public TestCase
{
public:
  SkipTest(const std::string& name, 
           const std::string& reason) :
    TestCase(name),
    reason_(reason)
  {
  } // SkipTest

protected:
  virtual void runTest()
  {
    throw SkipException(reason_);
  } // runTest

private:
  std::string reason_;
}; // class SkipTest

class CompileFailsTest : public TestCase
{
public:
  CompileFailsTest(const std::string& name,
                   const std::string& input_xslt) :
    TestCase(name),
    input_xslt_(input_xslt)
  {
  } // CompileFailsTest

protected:
  virtual void runTest()
  {
    Arabica::XSLT::StylesheetCompiler compiler;

    Arabica::SAX::InputSource source(input_xslt_);
    std::auto_ptr<Arabica::XSLT::Stylesheet> stylesheet = compiler.compile(source);
    if(stylesheet.get() != 0)
      assertImplementation(false, "Expected " + input_xslt_ + " not to compile.  But it did :o");
    throw SkipException(compiler.error());
  } // runTest

private:
  std::string input_xslt_;
}; // CompileFailsTest

class RunFailsTest : public TestCase
{
public:
  RunFailsTest(const std::string& name,
               const std::string& input_xml,
               const std::string& input_xslt) :
    TestCase(name),
    input_xml_(input_xml),
    input_xslt_(input_xslt)
  {
  } // RunFailsTest

protected:
  virtual void runTest()
  {
    Arabica::XSLT::StylesheetCompiler compiler;

    Arabica::SAX::InputSource source(input_xslt_);
    std::auto_ptr<Arabica::XSLT::Stylesheet> stylesheet = compiler.compile(source);
    if(stylesheet.get() == 0)
      assertImplementation(false, "Failed to compile " + input_xslt_ + " : " + compiler.error());

    Arabica::XSLT::DOMSink output;
    stylesheet->set_output(output);

    std::ostringstream errors;
    stylesheet->set_error_output(errors);

    Arabica::DOM::Document<std::string> document = buildDOM(input_xml_); 
    try {
      stylesheet->execute(document);
    }
    catch(const std::exception& e) {
      throw SkipException(e.what());
    }
    assertImplementation(false, "Expected " + input_xslt_ + " to fail to execute.  But it did :o");
  } // runTest

private:
  std::string input_xml_;
  std::string input_xslt_;
}; // RunFailsTest

class StandardXSLTTest : public TestCase
{
public:
  StandardXSLTTest(const std::string& name,
                   const std::string& input_xml,
                   const std::string& input_xslt,
                   const std::string& output_xml) : 
    TestCase(name),
    input_xml_(input_xml),
    input_xslt_(input_xslt),
    output_xml_(output_xml)
  { 
  } // StandardXSLTTest

protected:
  virtual void  runTest()
  {
    Arabica::XSLT::StylesheetCompiler compiler;

    Arabica::SAX::InputSource source(input_xslt_);
    std::auto_ptr<Arabica::XSLT::Stylesheet> stylesheet = compiler.compile(source);
    if(stylesheet.get() == 0)
      assertImplementation(false, "Failed to compile " + input_xslt_ + " : " + compiler.error());

    Arabica::XSLT::DOMSink output;
    stylesheet->set_output(output);

    std::ostringstream errors;
    stylesheet->set_error_output(errors);

    Arabica::DOM::Document<std::string> document = buildDOM(input_xml_); 
    try {
      stylesheet->execute(document);
    }
    catch(const std::exception& e) {
      assertImplementation(false, "Failed to run " + input_xslt_ + " : " + e.what());
    }

    Arabica::DOM::Document<std::string> ref = buildDOM(output_xml_);
    if(ref == 0)
      assertImplementation(false, "Couldn't read " + output_xml_ + ". Perhaps it isn't well-formed XML?");
    Arabica::DOM::Node<std::string> out = output.node();

    std::string refs = docToString(ref.getFirstChild());
    std::string outs = docToString(out.getFirstChild());

    if(refs != outs)
    {
      stripWhitespace(ref);
      stripWhitespace(out);
      std::string refs2 = docToString(ref.getFirstChild());
      std::string outs2 = docToString(out.getFirstChild());
      if(refs2 != outs2)
        assertImplementation(false, "Expected:\n" + refs + "\nbut got:\n" + outs);
    }
  } // runTest

  std::string docToString(Arabica::DOM::Node<std::string> node)
  {
    std::ostringstream ss;
    ss << node;
    return Arabica::string::normalize_whitespace<std::string, Arabica::default_string_adaptor<std::string> >(ss.str());
  } // docToString

  void stripWhitespace(Arabica::DOM::Node<std::string> doc)
  {
    Arabica::XPath::NodeSet<std::string> textNodes = selectNodes("//text()", doc);
    for(int i = 0; i != textNodes.size(); ++i)
    {
      Arabica::DOM::Node<std::string> t = textNodes[i];
      std::string text = t.getNodeValue();
      size_t index = text.find_first_of(" \n");
      while(index != std::string::npos)
      {
        text.replace(index, 1, "");
        index = text.find_first_of(" \n");
      }
      t.setNodeValue(text);
    }
  } // stripWhitespace

  std::string input_xml_;
  std::string input_xslt_;
  std::string output_xml_;
}; // class StandardXSLTTest

class Expected 
{
public:
  Expected() 
  {
    Arabica::DOM::Document<std::string> fail_doc = buildDOM(PATH_PREFIX + "arabica-expected-fails.xml");
    Arabica::XPath::NodeSet<std::string> failcases = selectNodes("/test-suite/test-case", fail_doc);
    for(int i = 0; i != failcases.size(); ++i)
    {
      std::string name = selectString("@id", failcases[i]);
      std::string compiles = selectString("@compiles", failcases[i]);
      std::string runs = selectString("@runs", failcases[i]);
      std::string skip = selectString("@skip", failcases[i]);
      std::string reason = selectString("@reason", failcases[i]);

      if(compiles == "no")
        fails[name] = "compile";
      else if(runs == "no")
        fails[name] = "run";
      else if(skip == "yes")
      {
        fails[name] = "skip";
        skips[name] = reason;
      }
    } // for ...
  } 

  std::map<std::string, std::string>& Fails() { return fails; }
  std::map<std::string, std::string>& Skips() { return skips; }

private:
  std::map<std::string, std::string> fails;
  std::map<std::string, std::string> skips;
};

static Expected expected;

TestSuite* XSLTTest_suite(const std::string& path)
{
  static Arabica::DOM::Document<std::string> catalog = buildDOM(PATH_PREFIX  + "catalog.xml");
  
  TestSuite *suiteOfTests = new TestSuite;

  Arabica::XPath::NodeSet<std::string> tests = 
//        selectNodes("/test-suite/test-catalog/test-case[scenario/@operation='standard']", catalog);
//        selectNodes("/test-suite/test-catalog/test-case[@id='boolean_boolean86']", catalog);
        selectNodes("/test-suite/test-catalog/test-case[scenario/@operation='standard' and file-path='" + path + "']", catalog);
  for(int i = 0; i != tests.size(); ++i)
  {
    std::string name = selectString("@id", tests[i]);
    std::string path = selectString("concat(../major-path, '" + SEPERATOR + "', file-path)", tests[i]);
    std::string out_path = selectString("concat(../major-path, '" + SEPERATOR + "REF_OUT" + SEPERATOR + "', file-path)", tests[i]);
    std::string input_xml = selectString(".//input-file[@role='principal-data']", tests[i]);
    std::string input_xslt = selectString(".//input-file[@role='principal-stylesheet']", tests[i]);
    std::string output_xml = selectString(".//output-file[@role='principal']", tests[i]);

    if(expected.Fails().find(name) == expected.Fails().end())
      suiteOfTests->addTest(new StandardXSLTTest(name, 
                                                 make_path(path, input_xml), 
                                                 make_path(path, input_xslt), 
                                                 make_path(out_path, output_xml)));
    else if(expected.Fails()[name] == "compile") 
      suiteOfTests->addTest(new CompileFailsTest(name, 
                                                 make_path(path, input_xslt)));
    else if(expected.Fails()[name] == "run")
      suiteOfTests->addTest(new RunFailsTest(name, 
                                             make_path(path, input_xml), 
                                             make_path(path, input_xslt)));
    else if(expected.Fails()[name] == "skip")
      suiteOfTests->addTest(new SkipTest(name, expected.Skips()[name]));
  } // for ...

	return suiteOfTests;
} // XSLTTest_suitoo

#endif
