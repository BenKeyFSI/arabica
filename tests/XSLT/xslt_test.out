
attribvaltemplate	12	0	0	1
axes	130	0	0	2
boolean	90	0	0	1
conditional	23	0	0	0
conflictres	35	0	0	1
copy	62	0	0	6
dflt	4	0	0	0
expression	6	0	0	6
extend	4	0	0	4
impincl	29	0	0	1
lre	22	2	0	1
match	32	0	0	2
math	107	2	0	0
mdocs	18	0	0	9
message	16	0	0	2
modes	17	0	0	0
namedtemplate	19	0	0	1
namespace	133	31	0	0
node	21	0	0	0
output	108	0	0	41
position	111	0	0	1
predicate	58	0	0	0
processorinfo	1	0	0	1
reluri	11	0	0	3
select	85	0	0	1
sort	37	3	0	10
string	133	4	0	8
variable	70	0	0	2
ver	5	0	0	4
whitespace	22	1	0	10
AVTs	25	2	0	0
Attributes	83	9	0	7
BVTs	80	15	0	19
Comment	17	0	0	4
Completeness	9	0	0	0
ConflictResolution	40	6	0	1
Copying	5	0	0	1
Elements	69	8	0	1
Errors	107	4	0	0
Fallback	10	0	0	7
ForEach	1	0	0	0
ForwardComp	18	0	0	13
Import	44	1	0	5
Keys	41	0	0	6
Messages	11	1	0	0
Miscellaneous	56	0	0	3
Modes	19	0	0	0
NamedTemplates	27	3	0	1
Namespace	43	1	0	4
Namespace-alias	40	20	0	0
Namespace_XPath	9	0	0	2
ProcessingInstruction	19	0	0	1
RTF	3	0	0	1
Sorting	124	5	0	27
Stylesheet	17	4	0	12
Template	17	0	0	0
Text	38	1	0	4
Valueof	4	0	0	0
Variables	83	0	0	10
Whitespaces	49	0	0	38
XSLTFunctions	97	2	0	34
attributes	1	0	0	0
errors	9	0	0	0
include	2	0	0	0
processing-instruction	1	0	0	0
stylesheet	3	0	0	0
text	1	0	0	0
variables	15	0	0	0
